core/bin/migrations/20240603_backfill_stored_documents.rs

```diff
@@ -0,0 +1,348 @@
+        data_source::{make_document_id_hash, DataSource, Section},
+        file_storage_document::FileStorageDocument,
+        qdrant::QdrantClients,
+    },
+    stores::{postgres, store},
+    utils,
+};
+use qdrant_client::qdrant;
+use tokio_postgres::Row;
+
+use bb8::Pool;
+use bb8_postgres::PostgresConnectionManager;
+use futures::{StreamExt, TryStreamExt};
+use tokio_postgres::NoTls;
+use tokio_stream::{self as stream};
+
+async fn fetch_data_sources_batch(
+    pool: &Pool<PostgresConnectionManager<NoTls>>,
+    last_id: u64,
+    limit: usize,
+) -> Result<Vec<Row>, anyhow::Error> {
+    let c = pool.get().await?;
+
+    c.query(
+        "SELECT id, internal_id FROM data_sources WHERE id > $1 LIMIT $2",
+        &[&(last_id as i64), &(limit as i64)],
+    )
+    .await
+    .context("Query execution failed")
+}
+
+// Return a "best-effort" section for the latest version of the document (based on Qdrant).
+// For superseded versions, return an empty section.
+async fn get_sections_from_document_version(
+    qdrant_clients: &QdrantClients,
+    ds: &DataSource,
+    document_id_hash: &str,
+    document_status: &str,
+    full_text: &str,
+) -> Result<Section> {
+    if document_status != "latest" {
+        return Ok(Section {
+            prefix: None,
+            content: Some(full_text.to_string()),
+            sections: vec![],
+        });
+    }
+
+    let qdrant_client = ds.main_qdrant_client(qdrant_clients);
+
+    let f = qdrant::Filter {
+        must: vec![qdrant::Condition::matches(
+            "document_id_hash",
+            document_id_hash.to_string(),
+        )],
+        ..Default::default()
+    };
+
+    let mut retry = 0;
+    let points_per_request = 256;
+    let mut page_offset: Option<PointId> = None;
+    let mut root_section = Section {
+        prefix: None,
+        content: None,
+        sections: vec![],
+    };
+
+    loop {
+        let scroll_results = match qdrant_client
+            .scroll(
+                &ds.embedder_config(),
+                &ds.internal_id().to_string(),
+                Some(f.clone()),
+                Some(points_per_request as u32),
+                page_offset.clone(),
+                Some(false.into()),
+            )
+            .await
+        {
+            Ok(r) => r,
+            Err(e) => {
+                if retry < 3 {
+                    retry += 1;
+                    utils::error(&format!(
+                        "Error migrating points (read): retry={} error={:?}",
+                        retry, e
+                    ));
+                    tokio::time::sleep(tokio::time::Duration::from_secs(1)).await;
+                    continue;
+                } else {
+                    Err(e)?
+                }
+            }
+        };
+
+        let mut sorted_results = scroll_results.result.clone();
+
+        // Sort results by chunk_offset. We don't have the proper index to sort by `chunk_offset`.
+        sorted_results.sort_by_key(|r| {
```

Comment:
Note I think qdrant supports sorting. But def fine to do it here :+1:

Reply:
Yes it does but we did not setup the right indexes.